### ISSUE TITLE
Error when loss contains scalar multiplication (eg. L2 regularization)

### ISSUE THREAD

**Issue by alex-weaver** [twitter/torch-autograd#73]
I ran across an issue trying to implement a sparsity-inducing term in the cost function, and in trying to put together a minimal repro case, it turned out that the example train-mnist-autoencoder.lua fails with the following output:

```
Training Epoch #1	
/home/alex/torch/install/bin/luajit: invalid arguments: number 
expected arguments: DoubleTensor | [*DoubleTensor*] DoubleTensor index
stack traceback:
	[C]: at 0x7f6b925e2940
	[C]: in function 'gf'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:58: in function 'gradOnly'
	...all/share/lua/5.1/autograd/runtime/direct/DirectTape.lua:124: in function 'df'
	train-mnist-autoencoder.lua:90: in main chunk
	[C]: in function 'dofile'
	...alex/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00406670
```
note to get this to run I had to remove the line
    `local Value = require 'autograd.Value'`
and change the following:
```lua
--for i=1,Value.len(params.W) do
for i=1,#params.W do
```
Hopefully these modifications don't break the example.

There seems to be an issue around the scalar multiplication of the summed square weights with the l2Lambda parameter. If the l2Lambda term is removed from the loss calculation, but the sum of squares of weights is left in, then the above error does not occur:
```lua
--loss = loss + l2Lambda * torch.sum(torch.pow(params.W[i],2))
  loss = loss + torch.sum(torch.pow(params.W[i],2))
```

The same behaviour happens both with and without optimization enabled. If I set a debugHook, it is never triggered (presumably because this issue doesn't involve `nan` or `inf`). I've also tried rebuilding torch with Lua 5.1 instead of LuaJIT, and a similar error occurs.

**Comment by alexbw** [comment]
You can't reuse a variable name, so you'd have to do something like `loss1 = loss2 + torch.sum(torch.pow(params.W[i],2))`

**Comment by alexbw** [comment]
Good catch, I'll figure out a fix. We shouldn't be re-using variable names.

**Comment by alex-weaver** [comment]
I'm not sure that the reuse of variable names is what is causing this particular issue (unless you're referring to something else?)

I've tried manually unrolling the loop in the mnist autoencoder example, renaming the loss variable each time, and the error still occurs if the l2Lambda term is left in.

To keep the failing case simple, including only weight decay on the first layer of weights while renaming the loss variable still triggers the error:
```lua
function f(params, input, l2Lambda)
   -- Reconstruction loss
   local prediction = predict(params, input)
   local loss = lossFuns.logBCELoss(prediction, input, 1e-6) / torch.size(input, 1)
   -- L2 penalty on the weights (NB: first layer only, just to test!)
   local loss2 = loss + l2Lambda * torch.sum(torch.pow(params.W[1],2))
   return loss2, prediction
end
```
Again, removing the l2Lambda term doesn't trigger the error.

**Comment by alexbw** [comment]
Interesting. Thanks for your work on this. Could you reduce this even further into a self-contained test case? That way, we can incorporate it into our unit tests.

**Comment by alex-weaver** [comment]
Sure, this is the smallest I can make an example that still fails:
```
local grad = require 'autograd'
local lossFuns = require 'autograd.loss'
local util = require 'autograd.util'

local trainData = torch.randn(10,5)

function f(params, input, l2Lambda)
	local loss = torch.sum(input * params.W)
	
	local loss2 = loss + torch.sum(params.W) * l2Lambda
	return loss2
end

local df = grad(f)

params = {
	W = torch.randn(5,5)
}

local l2LambdaValue = 0.4
local grads, loss = df(params,trainData,l2LambdaValue)
```

removing the multiply by l2Lambda on the 10th line, this code runs to completion, otherwise the call to df fails with
```
invalid arguments: number 
expected arguments: DoubleTensor | [*DoubleTensor*] DoubleTensor index
```

on a sidenote, if l2LambdaValue is left undefined (I did this accidentally while putting together this test case), it ends up failing with `C stack overflow`. I don't know how feasible it would be to catch this kind of mistake earlier to be able to provide a clearer error?

Anyway, I think that's as simple as I can make this.

**Comment by alex-weaver** [comment]
I've just taken another look at this, and I think I've found the issue. It looks like operators.mul in gradfuns.lua assumes that if one argument is not a tensor, then the other one is; with scalar-scalar multiply this is not the case. Changing the first condition slightly seems to fix it. I'll submit a pull request in a minute with the fix and a unit test

**Comment by luketwitter** [comment]
Thanks for the patch!
